primitives/src/asset_metadata.rs

```diff
@@ -0,0 +1,227 @@
+    pub fn is_locked(&self, now: Moment) -> bool {
+        self.lock_status.is_locked(now)
+    }
+}
+
+/// Asset Metadata Lock Status
+#[derive(Encode, Decode, TypeInfo)]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum AssetMetadataLockStatus<Moment> {
+    /// Can be changed by asset issuer.
+    Unlocked,
+    /// Cannot be changed.
+    Locked,
+    /// Cannot be changed until `Moment`.
+    LockedUntil(Moment),
+}
+
+impl<Moment: PartialOrd> AssetMetadataLockStatus<Moment> {
+    /// Check if the lock status is locked.
+    pub fn is_locked(&self, now: Moment) -> bool {
+        match self {
+            Self::Unlocked => false,
+            Self::Locked => true,
+            Self::LockedUntil(until) => now > *until,
+        }
+    }
+}
+
+impl<Moment> Default for AssetMetadataLockStatus<Moment> {
+    fn default() -> Self {
+        Self::Unlocked
+    }
+}
+
+/// Asset Metadata description.
+#[derive(Encode, Decode, TypeInfo, VecU8StrongTyped)]
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct AssetMetadataDescription(pub Vec<u8>);
+
+/// Asset Metadata Specs.
+#[derive(Encode, Decode, TypeInfo)]
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct AssetMetadataSpec {
+    /// Off-chain specs or documentation.
+    pub url: Option<Url>,
+    /// Description of metadata type.
+    pub description: Option<AssetMetadataDescription>,
+    /// Optional SCALE encoded `AssetMetadataTypeDef`.
+    pub type_def: Option<Vec<u8>>,
+}
+
+impl AssetMetadataSpec {
+    /// Returns the complexity of the metadata specs.
+    pub fn complexity(&self) -> usize {
+        let mut complexity = 0usize;
+        if self.url.is_some() {
+            complexity += 1;
+        }
+        if self.description.is_some() {
+            complexity += 1;
+        }
+        if self.type_def.is_some() {
+            complexity += 1;
+        }
+        complexity
+    }
+
+    /// Set the metadata type definition from an `AssetMetadataTypeDef`.
+    ///
+    /// This will encode the `AssetMetadataTypeDef` to a byte array.
+    pub fn set_type_def(&mut self, type_def: AssetMetadataTypeDef) {
+        self.type_def = Some(type_def.encode());
+    }
+
+    /// Decode the metadata type definition from `self.type_def` and return it as an `AssetMetadataTypeDef`.
+    pub fn decode_type_def(&self) -> Result<Option<AssetMetadataTypeDef>, codec::Error> {
+        self.type_def
+            .as_ref()
+            .map(|d| AssetMetadataTypeDef::decode_all(&d[..]))
+            .transpose()
+    }
+
+    /// Return the type definition length.
+    pub fn type_def_len(&self) -> usize {
+        self.type_def.as_ref().map(|d| d.len()).unwrap_or_default()
+    }
+}
+
+/// Asset Metadata type definition.
+#[derive(Encode, Decode)]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct AssetMetadataTypeDef {
+    /// A registry of type definitions.
+    pub types: PortableRegistry,
+    /// The top-level type id.
+    #[codec(compact)]
+    pub ty: u32,
```

Comment:
Forgot to ask why the `compact` is here... could you add a non-doc comment with the explanation?

Reply:
Because most of the time type ids will be small and only need 1 byte here.